Composer/packages/server/src/models/extension/composerEventEmitter.ts

```diff
@@ -0,0 +1,73 @@
+// Copyright (c) Microsoft Corporation.
```

Comment:
You can do the following if you want it typed with vscode intellisense:
```typescript
type EventTable = {
  eventName: (arg1: string, arg2: string) => void;
  myEvent: (num: number, cb: () => void) => void;
};

type EventName = keyof EventTable;

type ArgsOf<T> = T extends (...args: infer U) => void ? U : never;

type Disposable = { dispose: () => void };

class Emitter<T extends EventTable> {
  private events: Record<string, Record<string, Function>>;

  private uid: number;

  constructor() {
    this.events = {};
    this.uid = 0;
  }

  public on<TName extends keyof EventTable>(
    eventName: TName,
    callback: (...args: ArgsOf<T[TName]>) => void
  ): Disposable {
    if (!this.events[eventName]) {
      this.events[eventName] = {};
    }
    const cbId = this.uid++;
    this.events[eventName][cbId] = callback;
    return {
      dispose: () => {
        delete this.events[eventName][cbId];
        if (!Object.keys(this.events[eventName]).length) {
          delete this.events[eventName];
        }
      },
    };
  }

  public emit<TName extends keyof EventTable>(
    eventName: TName,
    ...args: ArgsOf<T[TName]>
  ) {
    if (!this.events[eventName]) {
      throw `event: ${eventName} does not exist!`;
    }
    for (const cbId in this.events[eventName]) {
      this.events[eventName][cbId].apply(null, args);
    }
  }
}

const emitter = new Emitter<EventTable>();
const sub = emitter.on("eventName", (arg1, arg2) =>
  console.log(arg1, arg2)
);
emitter.emit("eventName", "test", "test");

sub.dispose();
```

Reply:
This is cool. I'll make the change!